### ISSUE TITLE
Encoding Players require numpy

### ISSUE THREAD

**Issue by benzax** [chikinn/hanabi#40]
At least for me, this makes testing python3 compatibility a hassle.

Can we easily replace this usage?  Should I just install numpy and get over it?

**Comment by chikinn** [comment]
I was wondering about that too, I looked through Encoding Player and didn't see a quick fix.  But I also didn't immediately understand what numpy was being used for.

What do you think, Taylor?  @robieta

**Comment by robieta** [comment]
The simple answer is that it would take a major rewrite to remove the numpy dependency. Things that encoding player used numpy for:

1) Basic math functions (min, max, sum). Numpy is just really convenient for these because it is flexible in what the inputs can be, so coding is easy.
2) Arrays. Python doesn't natively have a good matrix analog. In the encoding AI I use arrays because they are an easy way to do element wise operations. (Yes I could do python-ey things like nested loops, but "np.array(A) \* np.array(B)" is easier.)
3) Unique. Having a function to get unique elements is convenient, and this way I don't have to write my own.
4) array_equal, and similar functions. This class of functions makes comparisons of lists much easier.

Bottom line: You pretty much never NEED numpy, but it is a really convenient and very versatile toolkit. (And I would be very very sad if I had to stop using it.)

**Comment by benzax** [comment]
Given the utility of having numpy for the encoding players, I think the best solution we're going to get is #41 .  People who don't have numpy installed just won't be able to use the encoding players.  Maybe if every other bot started using the same small subset of numpy it would make sense to write our own, but here clearly not.
